tools/setool/lib/create-module/generator.js

```diff
@@ -0,0 +1,922 @@
+	$(GSUTIL) cp -r ../${lname}_library.js $(BQ_BUCKET_PUBLIC)$(BQ_DATASET_${uname})/
+
+storage_remove: check_environment
+	$(GSUTIL) rm -rf $(BQ_BUCKET_PUBLIC)$(BQ_DATASET_${uname})/
+`)}
+##################### BIGQUERY DATASET #####################
+dataset_create: check_environment
+	$(BQ) --project_id $(BQ_PROJECTID) show $(BQ_DATASET_${uname}) 2>/dev/null 1>/dev/null || \\
+		$(BQ) mk -d --description "${uname} Dataset" $(BQ_PROJECTID):$(BQ_DATASET_${uname})
+
+dataset_remove: check_environment
+	$(BQ) rm -r -f -d $(BQ_PROJECTID):$(BQ_DATASET_${uname})
+
+REPLACEMENTS = -e 's!@@BQ_PROJECTID@@!$(BQ_PROJECTID)!g' \\
+	-e 's!@@BQ_DATASET_${uname}@@!$(BQ_DATASET_${uname})!g'${islib(` \\
+	-e 's!@@${uname}_BQ_LIBRARY@@!$(${uname}_BQ_LIBRARY)!g'`)}
+
+dataset_deploy: check_environment
+	for n in $(sort $(wildcard sql/*.sql)); do \\
+		$(SED) $(REPLACEMENTS) $$n | $(BQ) -q --project_id $(BQ_PROJECTID) query --use_legacy_sql=false || exit; \\
+	done
+
+##################### DEPLOY #####################
+deploy: check_environment${islib(`
+	$(MAKE) storage_upload`)}
+	$(MAKE) dataset_create
+	$(MAKE) dataset_deploy
+
+##################### INTEGRATION TESTS #####################
+check-integration: check_environment
+	$(MAKE) deploy
+	$(MAKE) -C test/ $@ || ($(MAKE) integration_cleanup && exit 1)
+	$(MAKE) integration_cleanup
+
+# Note, on failure we add a explicit sleep to wait until all resources are unused before retrying
+integration_cleanup: check_environment
+ifeq ($(POST_INTEGRATION_CLEANUP), 1)${islib(`
+	$(MAKE) storage_remove`)}
+	$(MAKE) dataset_remove || ((sleep 5 && $(MAKE) dataset_remove) || exit 1)
+endif`;
+
+    function islib (text) {
+        return library ? text : '';
+    }
+
+    createFile([name, 'bq', 'Makefile'], content);
+}
+
+function createBQDocReference (name, visibility) {
+    const lname = name.toLowerCase();
+    const badge = visibility === 'private' ? 'advanced' : 'core';
+    const content = `## ${lname}
+
+<div class="badge ${badge}"></div>
+
+TODO: module description.
+
+### VERSION
+
+{{% bannerNote type="code" %}}
+${lname}.VERSION()
+{{%/ bannerNote %}}
+
+**Description**
+
+Returns the current version of the ${lname} module.
+
+**Return type**
+
+\`STRING\`
+
+**Example**
+
+\`\`\`sql
+SELECT bqcarto.${lname}.VERSION();
+-- 1.0.0
+\`\`\``;
+
+    createFile([name, 'bq', 'doc', 'REFERENCE.md'], content);
+}
+
+function createBQSQLVersion (name, library) {
+    const lname = name.toLowerCase();
+    const uname = name.toUpperCase();
+    const content = `-----------------------------------------------------------------------
+--
+-- Copyright (C) 2021 CARTO
+--
+-----------------------------------------------------------------------
+
+CREATE OR REPLACE FUNCTION \`@@BQ_PROJECTID@@.@@BQ_DATASET_${uname}@@.VERSION\`()
+    RETURNS STRING
+    DETERMINISTIC
+    LANGUAGE js${islib(`
+    OPTIONS (library=["@@${uname}_BQ_LIBRARY@@"])`)}
+AS """
+    return ${library ? `${lname}Version()` : `'1.0.0'`};
```

Comment:
This looks good to me, but the code generated for the `constructor` module didn't have this right, have you already fixed it?

Reply:
I think the code in the `constructors` module is pure SQL, right?

Reply:
The code developed so far is pure SQL, however for the turf functions I'm pretty sure we are requiring JS code.